File: android/app/src/main/kotlin/com/pixeldev/Docly/pdf/PdfRenderView.kt
package com.pixeldev.Docly.pdf

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.Rect
import android.graphics.RectF
import android.view.View
import kotlin.math.abs

class PdfRenderView(
    context: Context,
    private val engine: PdfEngine,
) : View(context) {
    private val bitmapPaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.FILTER_BITMAP_FLAG)
    private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        strokeCap = Paint.Cap.ROUND
        strokeJoin = Paint.Join.ROUND
    }
    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.WHITE
        textSize = 18f
    }
    private val highlightPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
    }

    init {
        engine.attachInvalidator {
            postInvalidateOnAnimation()
        }
    }

    override fun onDetachedFromWindow() {
        engine.attachInvalidator(null)
        super.onDetachedFromWindow()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (w > 0 && h > 0) {
            engine.onSurfaceSizeChanged(w, h)
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawColor(Color.BLACK)

        val frame = engine.currentFrame()
        val base = frame.baseBitmap
        if (base == null) {
            drawLoading(canvas)
            return
        }

        canvas.drawBitmap(base, null, Rect(0, 0, width, height), bitmapPaint)
        drawHighlights(canvas, frame)
        drawStrokes(canvas, frame)
        drawTexts(canvas, frame)
        drawImages(canvas, frame)
    }

    private fun drawLoading(canvas: Canvas) {
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.WHITE
            textSize = 18f
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText("Loading PDF...", width / 2f, height / 2f, paint)
    }

    private fun drawHighlights(canvas: Canvas, frame: PdfRenderFrame) {
        frame.highlights.forEach { op ->
            highlightPaint.color = op.color
            highlightPaint.alpha = (255 * op.opacity).toInt().coerceIn(0, 255)
            canvas.drawRect(
                RectF(op.x.toFloat(), op.y.toFloat(), (op.x + op.w).toFloat(), (op.y + op.h).toFloat()),
                highlightPaint,
            )
        }
    }

    private fun drawStrokes(canvas: Canvas, frame: PdfRenderFrame) {
        frame.strokes.forEach { op ->
            if (op.points.size < 2) {
                return@forEach
            }

            strokePaint.color = op.color
            strokePaint.strokeWidth = op.width
            val path = Path()
            val first = op.points.first()
            path.moveTo(first.x, first.y)
            var previous = first
            for (index in 1 until op.points.size) {
                val point = op.points[index]
                val midX = (previous.x + point.x) / 2f
                val midY = (previous.y + point.y) / 2f
                path.quadTo(previous.x, previous.y, midX, midY)
                previous = point
            }
            path.lineTo(previous.x, previous.y)
            canvas.drawPath(path, strokePaint)
        }
    }

    private fun drawTexts(canvas: Canvas, frame: PdfRenderFrame) {
        frame.texts.forEach { op ->
            textPaint.color = op.color
            textPaint.textSize = op.fontSize.toFloat()
            canvas.drawText(op.text, op.x.toFloat(), op.y.toFloat(), textPaint)
        }
    }

    private fun drawImages(canvas: Canvas, frame: PdfRenderFrame) {
        frame.images.forEach { op ->
            val file = java.io.File(op.path)
            if (!file.exists()) {
                return@forEach
            }
            val bitmap = BitmapFactory.decodeFile(file.absolutePath) ?: return@forEach
            val left = op.x.toFloat()
            val top = op.y.toFloat()
            val right = left + op.width.toFloat().coerceAtLeast(1f)
            val bottom = top + op.height.toFloat().coerceAtLeast(1f)
            canvas.drawBitmap(bitmap, null, RectF(left, top, right, bottom), bitmapPaint)
        }
    }
}

File: ios/Runner/pdf/PdfRenderView.swift
import UIKit

final class PdfRenderView: UIView {
  private let engine: PdfEngine
  private let bitmapPaint = CGBlendMode.normal

  init(engine: PdfEngine) {
    self.engine = engine
    super.init(frame: .zero)
    backgroundColor = .black
    engine.attachInvalidator { [weak self] in
      self?.setNeedsDisplay()
    }
  }

  @available(*, unavailable)
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    if bounds.width > 0, bounds.height > 0 {
      engine.onSurfaceSizeChanged(width: Int(bounds.width), height: Int(bounds.height))
    }
  }

  override func draw(_ rect: CGRect) {
    guard let context = UIGraphicsGetCurrentContext() else { return }
    context.setFillColor(UIColor.black.cgColor)
    context.fill(bounds)

    let frame = engine.currentFrame()
    guard let baseImage = frame.baseImage else {
      drawLoading(in: context)
      return
    }

    baseImage.draw(in: bounds)
    drawHighlights(frame, in: context)
    drawStrokes(frame, in: context)
    drawTexts(frame, in: context)
    drawImages(frame, in: context)
  }

  private func drawLoading(in context: CGContext) {
    let paragraph = NSMutableParagraphStyle()
    paragraph.alignment = .center
    let attributes: [NSAttributedString.Key: Any] = [
      .foregroundColor: UIColor.white,
      .font: UIFont.systemFont(ofSize: 18, weight: .semibold),
      .paragraphStyle: paragraph,
    ]
    let text = NSAttributedString(string: "Loading PDF...", attributes: attributes)
    text.draw(in: CGRect(x: 0, y: bounds.midY - 16, width: bounds.width, height: 32))
  }

  private func drawHighlights(_ frame: PdfRenderFrame, in context: CGContext) {
    for op in frame.highlights {
      context.setFillColor(UIColor(red: CGFloat((op.color >> 16) & 0xFF) / 255.0,
                                   green: CGFloat((op.color >> 8) & 0xFF) / 255.0,
                                   blue: CGFloat(op.color & 0xFF) / 255.0,
                                   alpha: op.opacity).cgColor)
      context.fill(CGRect(x: op.x, y: op.y, width: op.w, height: op.h))
    }
  }

  private func drawStrokes(_ frame: PdfRenderFrame, in context: CGContext) {
    for op in frame.strokes where op.points.count > 1 {
      let color = UIColor(
        red: CGFloat((op.color >> 16) & 0xFF) / 255.0,
        green: CGFloat((op.color >> 8) & 0xFF) / 255.0,
        blue: CGFloat(op.color & 0xFF) / 255.0,
        alpha: CGFloat((op.color >> 24) & 0xFF) / 255.0
      )
      context.setStrokeColor(color.cgColor)
      context.setLineWidth(CGFloat(op.width))
      context.setLineCap(.round)
      context.setLineJoin(.round)
      context.beginPath()
      guard let first = op.points.first else { continue }
      context.move(to: CGPoint(x: first.x, y: first.y))
      var previous = first
      for point in op.points.dropFirst() {
        let mid = CGPoint(x: (previous.x + point.x) / 2.0, y: (previous.y + point.y) / 2.0)
        context.addQuadCurve(to: mid, control: CGPoint(x: previous.x, y: previous.y))
        previous = point
      }
      context.addLine(to: CGPoint(x: previous.x, y: previous.y))
      context.strokePath()
    }
  }

  private func drawTexts(_ frame: PdfRenderFrame, in context: CGContext) {
    for op in frame.texts {
      let color = UIColor(
        red: CGFloat((op.color >> 16) & 0xFF) / 255.0,
        green: CGFloat((op.color >> 8) & 0xFF) / 255.0,
        blue: CGFloat(op.color & 0xFF) / 255.0,
        alpha: CGFloat((op.color >> 24) & 0xFF) / 255.0
      )
      let paragraph = NSMutableParagraphStyle()
      paragraph.alignment = .left
      let attributes: [NSAttributedString.Key: Any] = [
        .foregroundColor: color,
        .font: UIFont.systemFont(ofSize: CGFloat(op.fontSize), weight: .regular),
        .paragraphStyle: paragraph,
      ]
      op.text.draw(at: CGPoint(x: op.x, y: op.y), withAttributes: attributes)
    }
  }

  private func drawImages(_ frame: PdfRenderFrame, in context: CGContext) {
    for op in frame.images {
      guard let image = UIImage(contentsOfFile: op.path) else { continue }
      image.draw(in: CGRect(x: op.x, y: op.y, width: max(op.width, 1), height: max(op.height, 1)))
    }
  }
}
